### ISSUE TITLE
Parameter issue

### ISSUE THREAD

**Issue by Tsuki0426** [JedMills/MTFL-For-Personalised-DNNs#9]
Hello JedMill,
I modified the parameters of lr and T when running fedavg. The reference parameters are set as follows. -dset mnist -alg fedavg -C 0.5 -B 20 -T 200 -E 1 -device gpu -W 200 -seed 0 -lr 0.01 -noisy_frac 0.0 -bn_private none, The test_acc with the largest running result is 0.9464. I would like to ask if there is a problem with this result. If there is a problem, what might be the problem?
Looking forward to your early reply.
Your sincerely,
Tang

**Comment by JedMills** [comment]
Hi Tang,

Thanks for your question. 

If you look in the file `Fig-5-settings.md`, it contains all the hyperparameters used to recreate Fig. 5 of the MTFL paper, as mentioned in the response to MiaJ321's issue. Looking at the relevant row for your settings:

| Dataset | W | C | Algorithm | Settings |
| -------- | -- | -- | --------- | --------- |
| MNIST | 200 | 0.5 | FL(FedAvg) | lr=0.3 | 

Shows that the learning rate used was lr=0.3, not lr=0.01. Can you try using lr=0.3 and see if your results corroborate with those in Fig. 5 (a)? I've also updated the readme with instructions for recreating the results.

Warm regards,
Jed

**Comment by Tsuki0426** [comment]
Hi Jed,

Thanks for your reply.

I used the lr=0.3 yestoday ，when the whole parameter are 

dset-mnist_alg-fedavg_C-0.5_B-20_T-500_E-1_device-gpu_W-200_seed-0_lr-0.3_noisy_frac-0.0_bn_private-non

and the results corroborate with those in the thesis.

However, I also have a question to know that in other papers, the result of fedavg is a relatively poor baseline in the case of non iid data division, which is far from our results. Is there any difference between them? Or is there something wrong with it?

Warm regards,
Tang

**Comment by JedMills** [comment]
Hi Tang,

That's good re. the hyperparameters. 

I'm not sure about the FedAvg performance. For example, looking at Fig. 2 of the original FedAvg [paper ](https://proceedings.mlr.press/v54/mcmahan17a/mcmahan17a.pdf) (using the same model architecture and non-IID MNIST partition as we did), their model reaches ~98% accuracy at around 200 rounds, which is close the the results we present in Fig. 5. Can you point me to the papers that have especially poor performance?

Warm regards,
Jed

**Comment by Tsuki0426** [comment]
Hello Jed,

Thanks for your reply.

I would like to re-state that what I was confused earlier by relatively poor baseline before is poorer performance relative to other algorithms at the same settings. Not in terms of a specific numerical magnitude, but in terms of the relative performance of the algorithms.

For example, in the articles 
PERSONALIZED FEDERATED LEARNING WITH FIRST ORDER MODEL OPTIMIZATION, 
and
Personalized Federated Learning with Moreau Envelopes,
the results of fedavg are less effective in the non-iid case relative to the other algorithms.

Warm regards,
Tang

**Comment by JedMills** [comment]
Hi Tang,

The difference in performance of FedAvg between the papers is likely due to differences in the experimental setups. For example, the pFedMe paper has only 20 clients using the same 2-class split. The FedFomo paper has 15 or 100 clients with a custom split. Each paper also uses different client-sampling rates and number of local epochs.

This represents a general problem in FL where authors arbitrarily partition existing datasets and run experiments on them. As FL is a new topic, authors were forced to adapt existing datasets in this way. However, attempts are being made to come up with standard benchmarks to allow better cross-comparison between papers, e.g. [LEAF](https://leaf.cmu.edu/). 

Verifying and cross-comparing results is generally very difficult, especially when authors do not publish their code (of a reasonable standard) or the hyperparameters used in their papers. 

Warm regards,
Jed

**Comment by Tsuki0426** [comment]
Hi Jed,

Thanks for your early reply.

I understand what you said, but maybe I didn't convey my meaning well. What I mean is that in the case that the general data is divided into non-iid, the effect of fedavg algorithm is generally lower than other algorithms under the **same parameter** configuration (excluding the specific parameters of the algorithm). 

But I wonder that I have recently reproduced the code in this article ,fedavg works better than many algorithms with the same parameters. For example, in the case of mnist no-iid, the effect of fedavg is generally better than that of perfedavg and pfedme when w=200 and w=400 under the same parameter configuration (excluding the specific parameters of the algorithm), same as the paper. Generally speaking, fedavg should be less effective than them, so I don't understand what the problem is.

Warm regards,
Tang

**Comment by Tsuki0426** [comment]
Hi Jed,
I'm sorry to bother you again. I was wondering if you had any comments on my previous question. Is the fedavg result normal or is there a problem? Do you see anything wrong with it?
I would appreciate it if you reply.
Warm regards,
Tang

**Comment by JedMills** [comment]
Hi Tang,

Apologies for the slow reply. 

In my experience, FedAvg generally works pretty well and is quite robust under a large variety of settings (compared with more complex algorithms), as shown by the experiments in our paper. 

Compared to the other personalised-FL algorithms using non-IID data, FedAvg seems to have pretty good performance. In fact, this is the main point of a recent [paper](https://arxiv.org/abs/2108.07313) which argues that FedAvg plus local fine-tuning has highly competitive performance compared to more sophisticated personalised-FL algorithms (e.g. see red curves in Figures 1-6). 

When proposing a new algorithm, authors are _highly_ incentivized to present it in a positive light (otherwise their papers will not be published). Therefore, many experiments are designed such that proposed algorithms perform very well and competing algorithms (i.e. FedAvg) very badly. Generally, I take the experimental results of papers that propose novel algorithms with a large pinch of salt. Surveys written by outside (disinterested) authors are usually more reliable.

Hope this clears up your concerns.

Warm regards,
Jed

**Comment by Tsuki0426** [comment]
Hi Jed,

Thanks for your reply. 

By your answers, my doubts are now relieved. Thank you for your kind help.

Warm regards,
Tang
